utils/vvflow/vvcompose.1.md
vvcompose(1) -- compose the CFD problem to be solved with vvflow program
====

## SYNOPSIS

`vvcompose` [<OPTIONS>] [--] [<FILE>] [<args>]

## DESCRIPTION

Moved to vvflow.1.md

## BODY GENERATORS

### gen_cylinder

    Moved to vvflow.1.md

### gen_semicyl

    Moved to vvflow.1.md

### gen_ellipse

    Moved to vvflow.1.md

### gen_plate

    Moved to vvflow.1.md

### gen_parallelogram

    Moved to vvflow.1.md

### gen_roundrect

    Moved to vvflow.1.md

### gen_chamber_gpj

    Moved to vvflow.1.md

### gen_chamber_box

    Moved to vvflow.1.md

### gen_savonius

    Moved to vvflow.1.md

## SEE ALSO
  vvflow(1), vvxtract(1), vvplot(1)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

- Merge `vvcompose` functionality with `vvflow`. The symlink is
  preserved for the compatibility.

## [2.3.0] - 2023-05-29

### Added

- `vvplot --eps` option for fine-tuning streamlines, vorticity and
  pressure fields smoothing factor.

### Fixed

- Buuilding instructions.
- Compilation errors with GCC 13.

## [2.2.2] - 2022-08-19

### Added

- Savonius shape generation for `vvcompose`.
- Binary packages for ubuntu jammy (22.04).

## [2.2.1] - 2021-02-18

### Fixed

- Upload packages for ubuntu focal and groovy.
- Revise the readme.

## [2.2.0] - 2020-11-12

### Added

- New vvplot option `-U` for plotting a velocity field.
- New vvplot option `--load-field` for plotting side-prcessed fields.

### Changed

- Replace `libmatheval` processor with Lua.
- Build (almost) all third-party libraries statically.
- Move CI from Bitbucket to the GitHub.
- Replace `googletest` testing framework with `cppunit`.

### Fixed

- Segmentation fault in vvplot due to stack-use-after-scope bug.

## [2.1.3] - 2020-02-26

### Fixed

- Fix the condition for using an inverse matrix when solving SLAE.
- Fix signs in Newton equation.
- Correct accounting of friction forces.

## [2.1.2] - 2020-01-09

### Added

- Support for negative spring damping coefficient.
- New vvplot `--ref-xy` option values: `bx`, `by`.

### Fixed

- Enhance vvencode video quality.

## [2.1.1] - 2019-02-03

### Fixed

- Mistake in external torque formula.

## [2.1.0] - 2019-01-22

### Fixed

- Correct formulae for bodies of zero area.

### Added

- Set external torque in holder.

## [2.0.0] - 2018-10-08

### Changed

- Introduce cmake-based build.
- Enable bitbucket-ci.
  - `googletest` unit tests.
  - `pytest` integration tests.
  - Publish deb packages.
- Include man pages.
- Rewrtite vvcompose in `C++`.
  - Built-in lua interpreter.
  - Built-in bodies generation.
- Rewrite vvxtract in `C++`.
  - Handle both results and stepdata.
- Rewrite vvplot in `C++`.
  - Save intermediate tar archives.


utils/vvcompose/vvcompose.cpp
#include "lua_space.h"
#include "lua_tvec.h"
#include "lua_tobj.h"
#include "lua_tvec3d.h"
#include "lua_tbody.h"
#include "lua_teval.h"
#include "lua_objlist.h"
#include "lua_bodylist.h"
#include "gen_body.h"
#include "getset.h"

#include "elementary.h"

#include <cstdio>
#include <cstring>
#include <unistd.h> // execlp
#include "linenoise.h"

int stackDump(lua_State *L)
{
    int i;
    int top = lua_gettop(L); /* depth of the stack */
    for (i = 1; i <= top; i++) { /* repeat for each level */
        int t = lua_type(L, i);
        switch (t) {
        case LUA_TSTRING: { /* strings */
            printf("#%d str  '%s'\n", i, lua_tostring(L, i));
            break;
        }
        case LUA_TBOOLEAN: { /* Booleans */
            printf("#%d bool %s\n", i, lua_toboolean(L, i) ? "true" : "false");
            break;
        }
        case LUA_TNUMBER: { /* numbers */
            printf("#%d num  %g\n", i, lua_tonumber(L, i));
            break;
        }
        case LUA_TUSERDATA: {
            const char *typ;
            if (luaL_getmetafield(L, i, "__name")) {
                typ = lua_tostring(L, -1);
                lua_pop(L, 1);
            }
            else
                typ = "userdata";
            printf("#%d udata %s\n", i, typ);
            break;
        }
        default: { /* other values */
            printf("#%d %s\n", i, lua_typename(L, t));
            break;
        }
        }
    }
    printf("\n"); /* end the listing */
    return 0;
}

int luaopen_vvd (lua_State *L) {

    lua_pushnumber(L, std::numeric_limits<double>::infinity()); // push 1
    lua_setglobal(L, "inf"); // pop 1
    lua_pushnumber(L, std::numeric_limits<double>::quiet_NaN()); // push 1
    lua_setglobal(L, "nan"); // pop 1

    lua_pushcfunction(L, luavvd_load_body); // push 1
    lua_setglobal(L, "load_body"); // pop 1
    lua_pushcfunction(L, luavvd_gen_cylinder); // push 1
    lua_setglobal(L, "gen_cylinder"); // pop 1
    lua_pushcfunction(L, luavvd_gen_semicyl); // push 1
    lua_setglobal(L, "gen_semicyl"); // pop 1
    lua_pushcfunction(L, luavvd_gen_ellipse); // push 1
    lua_setglobal(L, "gen_ellipse"); // pop 1
    lua_pushcfunction(L, luavvd_gen_plate); // push 1
    lua_setglobal(L, "gen_plate"); // pop 1
    lua_pushcfunction(L, luavvd_gen_parallelogram); // push 1
    lua_setglobal(L, "gen_parallelogram"); // pop 1
    lua_pushcfunction(L, luavvd_gen_chamber_gpj); // push 1
    lua_setglobal(L, "gen_chamber_gpj"); // pop 1
    lua_pushcfunction(L, luavvd_gen_chamber_box); // push 1
    lua_setglobal(L, "gen_chamber_box"); // pop 1
    lua_pushcfunction(L, luavvd_gen_savonius); // push 1
    lua_setglobal(L, "gen_savonius"); // pop 1

    luaopen_space(L);
    luaopen_tvec(L);
    luaopen_tobj(L);
    luaopen_tvec3d(L);
    luaopen_teval(L);
    luaopen_tbody(L);
    luaopen_objlist(L);
    luaopen_bodylist(L);

    return 0;
}

int main (int argc, char** argv) {

    if (argc > 1) {
        // try to parse arguments
        if (!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help")) {
            return execlp("man", "man", "vvcompose", NULL);
        } else if (!strcmp(argv[1], "-v") || !strcmp(argv[1], "--version")) {
            fprintf(stderr, "vvcompose %s\n", libvvhd_gitinfo);
            fprintf(stderr, "revision: %s\n", libvvhd_gitrev);
            if (libvvhd_gitdiff[0] == '\0') return 0;
            fprintf(stderr, "git_diff: %s\n", libvvhd_gitdiff);
            return 0;
        } else if (!strcmp(argv[1], "--")) {
            argv[1] = argv[0];
            argc--;
            argv++;
        } else if (!strncmp(argv[1], "-", 1)) {
            fprintf(stderr, "%s: invalid option '%s'\n", argv[0], argv[1]);
            return 1;
        }
    }

    lua_State *L = luaL_newstate();
    luaL_openlibs(L);
    luaopen_vvd(L);
    int ret = 0;

    if (argc > 1) {
        /* load script from file */
        lua_newtable(L);
        for (int i=1; i<argc; i++) {
            lua_pushstring(L, argv[i]);
            lua_rawseti(L, -2, i-1);
        }
        lua_setglobal(L, "arg");

        ret = luaL_dofile(L, argv[1]);

        if (ret) {
            fprintf(stderr, "%s\n", lua_tostring(L, -1));
            lua_pop(L, 1);
        }
    } else {
        /* interpreter mode */
        printf("vvcompose %s\n", libvvhd_gitinfo);
        printf("Press Ctrl+D to exit\n");
        char *line;
        while(true) {
            line = linenoise("vvcompose> ");
            if (errno == EAGAIN) {
                errno = 0;
                continue;
            } else if (!line) {
                break;
            }

            linenoiseHistoryAdd(line);

            int err = luaL_loadstring(L, line) || lua_pcall(L, 0, 0, 0);
            if (err) {
                printf("%s\n", lua_tostring(L, -1));
                lua_pop(L, 1);
            }

            linenoiseFree(line);
        }
    }

    lua_close(L);
    return ret;
}
